### ISSUE TITLE
Compiler assumes static_parameter in value position is effect-free

### ISSUE THREAD

**Issue by Keno** [JuliaLang/julia#45490]
```
julia> foo(x::Union{T, Nothing}, y::Union{T, Nothing}) where {T} = T
foo (generic function with 1 method)

julia> foo(nothing, nothing)
ERROR: UndefVarError: T not defined
Stacktrace:
 [1] foo(x::Nothing, y::Nothing)
   @ Main ./REPL[3]:1
 [2] top-level scope
   @ REPL[4]:1

julia> bar(x::Union{T, Nothing}, y::Union{T, Nothing}) where {T} = (typeof(T); nothing)
foo (generic function with 1 method)

julia> bar(nothing, nothing)
```

I thought we had already made the change to disallow static_parameter in value position, but apparently we only talked about it, but never actually changed the lowering. Fortunately, static_parameter with side effect is pretty rare, but nevertheless, this is a pretty ugly bug. We should change the lowering to disallow :static_parameter in value position.

**Comment by Keno** [comment]
I can fix this in IR conversion, but @JeffBezanson could you see if we can just do this in lowering? I had hoped
```
diff --git a/src/julia-syntax.scm b/src/julia-syntax.scm
index 4024d25c2e..ad072f05e9 100644
--- a/src/julia-syntax.scm
+++ b/src/julia-syntax.scm
@@ -4221,7 +4221,7 @@ f(x) = yt(x)
   (or (simple-atom? e) (symbol? e)
       (and (pair? e)
            (memq (car e) '(quote inert top core globalref outerref
-                                 slot static_parameter boundscheck)))))
+                                 slot boundscheck)))))
 
 (define (valid-ir-rvalue? lhs e)
   (or (ssavalue? lhs)
```
would be sufficient, but unfortunately, `:static_parameter` also gets introduced by `renumber-lambda`, which runs after linearize, so this is a bit more involved. If we do it during IR conversion, we further diverge what's legal in the IR and in CodeInfo, which seems suboptimal.

**Comment by Keno** [comment]
Fixed by the IR conversion change in #45459, but we should see if we can do the lowering change also.
